core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -308,6 +312,53 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     gcsDAO.deleteProject(projectName).map(_ => RequestComplete(StatusCodes.OK))
   }
 
+  def registerBillingProject(projectName: RawlsBillingProjectName, owner: RawlsUserEmail): Future[PerRequestMessage] = {
+    val cromwellAuthBucketName = s"cromwell-auth-${projectName.value}"
+
+    for {
+      _ <- dataSource.inTransaction { dataAccess =>
+        dataAccess.rawlsBillingProjectQuery.load(projectName) flatMap {
+          case None =>
+            for {
+              resource <- DBIO.from(samDAO.createResource(SamResourceTypeNames.billingProject, projectName.value, userInfo))
+              workspaceCreatorPolicy <- DBIO.from(samDAO.overwritePolicy(SamResourceTypeNames.billingProject, projectName.value, workspaceCreatorPolicyName, SamPolicy(Seq.empty, Seq.empty, Seq(SamProjectRoles.workspaceCreator)), userInfo))
+              canComputeUserPolicy <- DBIO.from(samDAO.overwritePolicy(SamResourceTypeNames.billingProject, projectName.value, canComputeUserPolicyName, SamPolicy(Seq.empty, Seq.empty, Seq(SamProjectRoles.batchComputeUser, SamProjectRoles.notebookUser)), userInfo))
+              groupEmail <- DBIO.from(samDAO.syncPolicyToGoogle(SamResourceTypeNames.billingProject, projectName.value, SamProjectRoles.owner, userInfo)).map(_.keys.headOption.getOrElse(throw new RawlsException("Error getting owner policy email")))
+              project <- dataAccess.rawlsBillingProjectQuery.create(
+                RawlsBillingProject(
+                  projectName,
+                  RawlsGroup(
+                    RawlsGroupName(SamProjectRoles.owner),
+                    groupEmail,
+                    Set.empty,
+                    Set.empty),
+                  cromwellAuthBucketName,
+                  CreationStatuses.Ready,
+                  None, // TODO: Do we need to enter the billing account name?
+                  None))
+            } yield project
+
+          case Some(_) => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, "project by that name already exists"))
```

Comment:
I would rewrite this a little bit. You have an outer for comprehension with 1 element in it. The inner for is mostly DBIO.froms so they can be mixed with dataAccess.rawlsBillingProjectQuery.create and all fit in an overall transaction. But we don't really want all these external system calls to happen within the transaction.

I would have 1 for comp that deals with Futures. You might even start off with creating the billing project in the db whcih will fail if it already exists.

```
for {
  project <- dataSource.inTransaction { dataAccess => dataAccess.rawlsBillingProjectQuery.create(...) }
  _ <- samDAO.createResource
  _ <- samDAO.overwritePolicy
...
} yield { project }